nwastdlib/asyncio_cache.py

```diff
@@ -0,0 +1,193 @@
+    provide your own serializer. In that case you can also disable the hmac signing by providing `secret=None`.
+
+    Note: Serialization errors will raise an Exception but deserialization will fail silently with only a logged error.
+
+
+    Example:
+        Decorate a suitable function with all the default settings like this:::
+            @cached_result(pool, "your_app_name", "SECRET_KEY_FOR_HMAC_CHECKSUM")
+            def my_cached_function(uuid_arg, string_arg, int_kwarg=1):
+                return do_stuff(uuid_arg, string_arg, kwarg)
+
+        The first call is cached and reused for 120s. If the defaults are inadequate use:::
+            @cached_result(pool, "your_app_name", "SECRET_KEY_FOR_HMAC_CHECKSUM", "desired_cache_key_name", 1800)
+            def my_other_function...
+
+    Args:
+        pool: A async redis cache pool.
+        secret: used to generate a checksum too ensure cache tampering is not possible. Set to `None` to disable.
+        prefix: Prefix for the cache keys generated.
+        key_name: When specified the total redis key_name will be "prefix:key_name".
+        expiry_seconds: expiration in seconds. Defaults to two minutes (120s).
+        serializer: Provide your own Serializer class or use the default pickle.
+
+    Returns:
+        decorator function
+
+    """
+
+    def cache_decorator(func: Callable) -> Callable:
+        @wraps(func)
+        async def func_wrapper(*args: tuple[Any], **kwargs: dict[str, Any]) -> Any:
+            python_major, python_minor = sys.version_info[:2]
+            if key_name:
+                cache_key = f"{prefix}:{python_major}.{python_minor}:{key_name}"
+            else:
+                # Auto generate a cache key name based on function_name and a hash of the arguments
+                # Note: this makes no attempt to handle non-hashable values like lists and sets or other complex objects
+                args_and_kwargs_string = (args, frozenset(kwargs.items()))
+                cache_key = f"{prefix}:{python_major}.{python_minor}:{func.__name__}{args_and_kwargs_string}"
+                logger.info("Autogenerated a cache key", cache_key=cache_key)
+
+            logger.debug("Cache called with wrapper func", func_name=func.__name__, cache_key=cache_key)
+            if secret:
+                if (cached_val := await get_signed_cache_value(pool, secret, cache_key, serializer)) is not None:
+                    logger.info("Cache contains secure key, serving from cache", cache_key=cache_key)
+                    return cached_val
+            else:
+                if (cached_val := await get_cache_value(pool, cache_key, serializer)) is not None:
+                    logger.info("Cache contains key, serving from cache", cache_key=cache_key)
+                    return cached_val
+
+            logger.info("Cache doesn't contain key, calling real function", cache_key=cache_key)
+            result = await func(*args, **kwargs)
```

Comment:
I don't think these log statements have any use in production, let's put them to `debug`?